ovn-tester/cms/openstack/openstack.py

```diff
@@ -0,0 +1,724 @@
+            localnet=True,
+        )
+        self.nbctl.ls_port_set_set_type(port, "localnet")
+        self.nbctl.ls_port_set_set_options(port, options)
+
+        return port
+
+    def _add_router_port_internal(
+        self, neutron_net: NeutronNetwork, router: LRouter, project: Project
+    ) -> (LSPort, LRPort):
+        """Add a pair of ports that connect router to the internal network.
+
+        The Router Port is automatically assigned network's gateway IP.
+
+        :param neutron_net: Neutron network that represents internal network
+        :param router: Router to which the network is connected
+        :param project: Project to which 'neutron_net' belongs
+        :return: The pair of Logical Switch Port and Logical Router Port
+        """
+        port_ip = DualStackIP(
+            neutron_net.gateway, neutron_net.network.cidr.prefixlen, None, None
+        )
+        return self._add_router_port(
+            neutron_net.network, router, port_ip, project.uuid, False
+        )
+
+    def _add_router_port_external_gw(
+        self, neutron_net: NeutronNetwork, router: LRouter
+    ) -> (LSPort, LRPort):
+        """Add a pair of ports that connect router to the external network.
+
+        The Router Port is assigned IP form external network's range.
+
+        :param neutron_net: Neutron network that represents external network
+        :param router: Router to which the network is connected
+        :return: The pair of Logical Switch Port and Logical Router Port
+        """
+        port_ip = self.next_external_ip()
+        return self._add_router_port(
+            neutron_net.network, router, port_ip, "", True
+        )
+
+    def _add_router_port(
+        self,
+        network: LSwitch,
+        router: LRouter,
+        port_ip: DualStackIP,
+        project_id: str = "",
+        is_gw: bool = False,
+    ) -> (LSPort, LRPort):
+        """Add a pair of ports that connect Logical Router and Logical Switch.
+
+        :param network: Logical Switch to which a Logical Switch Port is added
+        :param router: Logical Router to which a Logical Router Port is added
+        :param port_ip: IP assigned to the router port
+        :param project_id: Optional Openstack Project ID that's associated with
+                           the network.
+        :param is_gw: Set to True if Router port is connected to external
+                      network and should act as gateway.
+        :return: The pair of Logical Switch Port and Logical Router Port
+        """
+        subnet_id = str(uuid.uuid4())
+        port_name = uuid.uuid4()
+        router_port_name = f"lrp-{port_name}"
+        device_owner = "network:router_gateway" if is_gw else ""
+        # XXX: neutron adds "neutron-" prefix to router name (which is UUID),
+        #      but for the ports' external IDs we want to extract just the
+        #      UUID.
+        router_id = router.name.lstrip("neutron-")
+
+        lsp_external_ids = {
```

Comment:
(As a matter to discuss) A lot of effort here (passing parameters with "real" values, creating the dict) is spent to populate ext_ids with keys that are identical to a "real" openstack. But these keys are opaque for OVN, so we could shave off some complexity by either a) omitting populating the keys at all or b) if we want to make sure the db is populated with something, maybe just generate random strings / hardcode.

Reply:
We did ask ourselves whether it was worth populating them, and we concluded with three reasons to keep them:
1. To our surprise we found they are not entirely opaque, and there is actually processing of them in OVN [0][1].
2. Neutron's widespread use of external_ids increases the size of the database, we ought to measure what impact this has.
3. If you are developing/debugging the ovn-heater OpenStack plugin with a Neutron background, having these familiar keys with data that actually link to something can be useful.

0: https://github.com/ovn-org/ovn/blob/430c0fd32dd709b8d452ce9a339dd30228744b62/northd/northd.c#L1114-L1120
1: https://github.com/ovn-org/ovn/blob/430c0fd32dd709b8d452ce9a339dd30228744b62/northd/northd.c#L3683-L3690